backend/auth/role_map_repository.go

```diff
@@ -58,99 +59,221 @@ func getOrDefaultEnv(key, defaultValue string) string {
+	}
+	if _, hasResource := matrix["*"][deny.Resource]; !hasResource {
+		expandResources(deny.Resource, matrix)
+	}
 
-				if hasPermission(role, subroleMap, &op, make(map[string]struct{})) {
-					result[namespace][resource] = append(result[namespace][resource], opType)
+	if deny.Namespace == "*" {
+		if deny.Resource == "*" { // * * op
+			for _, resources := range matrix {
+				for _, operations := range resources {
+					removeOp(operations, deny.Type)
 				}
 			}
+		} else { // * r op
+			for _, resources := range matrix {
+				removeOp(resources[deny.Resource], deny.Type)
+			}
+		}
+	} else {
+		if deny.Resource == "*" { // n * op
+			for _, operations := range matrix[deny.Namespace] {
+				removeOp(operations, deny.Type)
+			}
+		} else { // n r op
+			removeOp(matrix[deny.Namespace][deny.Resource], deny.Type)
 		}
 	}
-
-	return result
 }
 
-type getFieldFromOp func(*models.Operation) string
-
-func findUsedOperationsInGraph(
-	role *models.Role,
-	subroleMap map[string]*models.Role, 
-	fieldFunction getFieldFromOp, 
-	values map[string]struct{}, 
-	visited map[string]struct{},
-) map[string]struct{}{
-	for _, op := range role.Deny {
-		value := fieldFunction(&op)
-		values[value] = struct{}{}
+func addOp(ops map[models.OperationType]struct{}, opType models.OperationType) {
+		if opType == models.All {
+			ops[models.Create] = struct{}{}
+			ops[models.Read] = struct{}{}
+			ops[models.Update] = struct{}{}
+			ops[models.Delete] = struct{}{}
+			ops[models.List] = struct{}{}
+		} else {
+			ops[opType] = struct{}{}
+		}
 	}
 
-	for _, op := range role.Permit {
-		value := fieldFunction(&op)
-		values[value] = struct{}{}
+func removeOp(ops map[models.OperationType]struct{}, opType models.OperationType) {
+	if opType == models.All {
+		delete(ops, models.Create)
+		delete(ops, models.Read)
+		delete(ops, models.Update)
+		delete(ops, models.Delete)
+		delete(ops, models.List)
+	} else {
+		delete(ops, opType)
 	}
+}
 
-	for _, subroleName := range role.Subroles {
-		subrole := subroleMap[subroleName]
-		findUsedOperationsInGraph(subrole, subroleMap, fieldFunction, values, visited)
+func expandNamespaces(namespace string, matrix map[string]map[string]map[models.OperationType]struct{}) {
+	matrix[namespace] = make(map[string]map[models.OperationType]struct{})
+	for resource, ops := range matrix["*"] {
+		matrix[namespace][resource] = make(map[models.OperationType]struct{})
+		for opType := range ops {
+			matrix[namespace][resource][opType] = struct{}{}
+		}
 	}
+}
 
-	return values
+func expandResources(resource string, matrix map[string]map[string]map[models.OperationType]struct{}) {
+	for _, namespaced := range matrix {
+		namespaced[resource] = make(map[models.OperationType]struct{})
+		for opType := range namespaced["*"] {
+			namespaced[resource][opType] = struct{}{}
+		}
+	}
 }
-	
 
 func hasPermission(role *models.Role, subroleMap map[string]*models.Role, operation *models.Operation, visited map[string]struct{}) bool {
```

Comment:
I need it for comparison tests atleast. Was thinking of using it as fallback if initializing the matrix would take a longer time then expected to, but honestly propably won't do that. Something about moving it to the test file doesn't sit right in me, but if you insist it could be done

Reply:
If it is a function used only in tests, then it should be in the test file, because now it looks like it is not used anywhere (for tests I just thought it is used in its tests, not in their creation), plus it spoils the test coverage, as if someone would ever look at it more